src/storage-types/src/connections.rs

```diff
@@ -1383,3 +1421,191 @@ impl AwsPrivatelinkConnection {
+                        self.inner.add_broker_rewrite(addr.clone(), move || {
+                            let addr = ssh_tunnel.local_addr();
+                            BrokerRewrite {
+                                host: addr.ip().to_string(),
+                                port: Some(addr.port()),
+                            }
+                        });
+
+                        let ret = self.inner.rewrite_broker_addr(addr);
+                        ret
+                    }
+                    Err(e) => {
+                        // Record the error...
+                        *self.error.lock().expect("poisoned") = Some(e.to_string());
+                        // ...and also ensure a log line shows up.
+                        self.error(KafkaError::ClientCreation(e.to_string()), &e.to_string());
+                        // We have to give rdkafka an address, as this callback can't fail,
+                        // so we have no better choice than the untranslated one.
+                        addr
+                    }
+                }
+            }
+            None => addr,
+        }
+    }
+
+    fn log(&self, level: RDKafkaLogLevel, fac: &str, log_message: &str) {
+        self.inner().log(level, fac, log_message)
+    }
+
+    fn error(&self, error: KafkaError, reason: &str) {
+        self.inner().error(error, reason)
+    }
+
+    fn stats(&self, statistics: Statistics) {
+        self.inner().stats(statistics)
+    }
+
+    fn stats_raw(&self, statistics: &[u8]) {
+        self.inner().stats_raw(statistics)
+    }
+
+    fn generate_oauth_token(
+        &self,
+        oauthbearer_config: Option<&str>,
+    ) -> Result<OAuthToken, Box<dyn std::error::Error>> {
+        self.inner().generate_oauth_token(oauthbearer_config)
+    }
+}
+
+impl<C> ConsumerContext for DynamicBrokerRewritingClientContext<C>
+where
+    C: ConsumerContext,
+{
+    fn rebalance(
+        &self,
+        native_client: &NativeClient,
+        err: RDKafkaRespErr,
+        tpl: &mut TopicPartitionList,
+    ) {
+        self.inner().rebalance(native_client, err, tpl)
+    }
+
+    fn pre_rebalance<'a>(&self, rebalance: &Rebalance<'a>) {
+        self.inner().pre_rebalance(rebalance)
+    }
+
+    fn post_rebalance<'a>(&self, rebalance: &Rebalance<'a>) {
+        self.inner().post_rebalance(rebalance)
+    }
+
+    fn commit_callback(&self, result: KafkaResult<()>, offsets: &TopicPartitionList) {
+        self.inner().commit_callback(result, offsets)
+    }
+
+    fn main_queue_min_poll_interval(&self) -> Timeout {
+        self.inner().main_queue_min_poll_interval()
+    }
+}
+
+impl<C> ProducerContext for DynamicBrokerRewritingClientContext<C>
+where
+    C: ProducerContext,
+{
+    type DeliveryOpaque = C::DeliveryOpaque;
+
+    fn delivery(
+        &self,
+        delivery_result: &DeliveryResult<'_>,
+        delivery_opaque: Self::DeliveryOpaque,
+    ) {
+        self.inner().delivery(delivery_result, delivery_opaque)
+    }
+}
```

Comment:
A number of comments on this from the original PR that are still outstanding: https://github.com/MaterializeInc/materialize/pull/18454#pullrequestreview-1362139531

The big one is that I think this should live in `kafka-util` and ideally be merged with the `BrokerRewritingClientContext`, so that we don't need the double wrapping and all the boilerplate that entails. I think that's now possible since you've hoisted the SSH tunnel manager out of `mz-storage` and into `mz-ssh-util`.

Reply:
done in #22795